File: 프로그래머스/lv2/12946. 하노이의 탑/하노이의 탑.py
answer = []
def hanoi(num, frm, to, other):
    global answer
    if num == 0:
        return
    hanoi(num-1, frm, other, to)
    answer.append([frm, to])
    hanoi(num - 1, other, to, frm)

def solution(n):
    frm = 1
    to = 3
    other = 2
    hanoi(n, frm, to, other)
    return answer

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int[][] solution(int n) {
        List<int[]> answerList = new ArrayList<>();

        List<int[]> result = hanoi(n, 1, 3, 2);

        return result.stream().toArray(int[][]::new);
    }
    
    public List<int[]> hanoi(int n, int from, int to, int other) {
        if (n == 0) return null;
        if (n == 1) {
            List<int[]> result = new ArrayList<>();
            result.add(new int[]{from, to});
            return result;
        }

        List<int[]> result = hanoi(n - 1, from, other, to);
        result.add(new int[]{from, to});
        result.addAll(hanoi(n - 1, other, to, from));
        return result;
    }
}